### PR TITLE
Update dapr devcontainer to v0.1.3 with Codespaces support

### LINKED ISSUE #2483: Devcontainer: Shared docker or docker-in-docker?
## In what area(s)?

/area runtime
> /area operator
> /area placement
> /area docs
> /area test-and-release

## Ask your question here
I saw that the devcontainer templates for dapr were recently moved to be hosted in the official repo at https://github.com/microsoft/vscode-dev-containers but I had a question around dapr and devcontainers in general...

I only started touching dapr in the last week or so, and spent a considerable amount of time getting stuck trying to run the intro/samples locally inside devcontainers (because that's how I partition my projects).  I'm honestly still not sure I have a handle on running it right; most attempts on the most basic example inside the dapr-node/typescript or a generic shared-docker devcontainer still give me errors at some point or another.

From what I can see, when "installing" dapr in standalone (non-k8s) mode, the expectation is for there to only be a single dapr ecosystem - binaries, config and containers - on a given host.  This makes sense for dedicated setups which use vanilla docker with dapr standalone.  But on a development machine it makes things difficult, as I might have different projects which aren't meant to co-exist within the same dapr environment.  

As such, it would seem to me intuitively that when running inside a devcontainer (where the design pattern of devcontainers is to fully host a given "project" in each devconainer) that sharing the host docker socket doesn't make sense.  A vanilla `dapr init` in devcontainer A would potentially conflict with a `dapr init` in devcontainer B because of container names (`dapr_placement`, `dapr_zipkin`, perhaps `dapr_redis`?) in a best case, and different configs in ~/.dapr in a worse case... 

The current devcontainer setup seeks to mitigate conflicts between the devcontainer and host, at least, by using docker-compose namespaces, but the devcontainer would still conflict with another devcontainer without modifications.

My question is, given the above, if it doesn't make sense to partition development environments with docker-in-docker setups rather than sharing the host docker socket?  Has this been discussed or explored in the past?

Thanks!

### LINKED ISSUE #3229: Support GitHub Codespaces for contributing to dapr project
## In what area(s)?

/area test-and-release

## Describe the proposal

The dapr project already has a VSCode devcontainer definition, but it does not load successfully as a [GitHub Codespace](https://github.com/features/codespaces) today. Providing a containerized dev environment with all of the appropriate prerequisites and tools for building and testing contributions to dapr would help reduce the friction in contributing to the process. For example, refer to @DarqueWarrior's post on [developing Dapr components in a container](https://www.donovanbrown.com/post/Develop-Dapr-Components-in-a-container).

The proposal is to update the devcontainer.json and Dockerfile-dev definitions with:

* Any necessary changes to make it compatible with Codespaces
* More general defaults that are likely to load successfully as local devcontainers without having to customize the devcontainer.json
   * Fix up the existing behaviors so there's no loss of functionality such as local docker binding, but make their behaviors more robust and explicitly opt-in.
   * This also means using the docker-in-docker configuration by default, as suggested by #2483

Specifically, the proposed composition of the container image would be:

* Based on golang-1.16 image (same as current devcontainer)
* VSCode-defined setup scripts to install:
   * Debian common features for Codespaces: sets up non-root user, access groups, interactive debian packages, etc.
   * Docker-in-docker, including docker group for access to /var/run/docker.sock
   * Kubectl, Helm and Minikube packages
* Reduced set of Go tools:
   * Remove deprecated tools and tools superseded by gopls currently defined in the devcontainer.
* Protoc compiler installed from binary instead of full source tree build to reduce image footprint.
* Add Dapr CLI for convenience
* Set Default Git editor to VSCode (instead of nano/vim) for better user experience in Codespace.
* Add a custom first-run-notice for dapr Codespace.

### PR DESCRIPTION
**Pull request by CodeMonkeyLeet**
# Description

Implement the devcontainer changes in proposal #3229, which include:
- Version the dapr-dev container image to v0.1.3
- Streamlined footprint of go tools and protoc install in daprdev container image.
- Add support for GitHub codespaces
- Change devcontainer default to use docker-in-docker configuration
- Remove direct bind-mount of host socket to /var/run/docker.sock in devcontainer which can cause ownership issues on Linux.
   - Disable binding to host network by default, but requires running devcontainer as `--privileged`. 
   - Add `BIND_LOCALHOST_DOCKER` option for user to restore redirect of /var/run/docker.sock to host socket as non-default setting. 
- Fix `SYNC_LOCALHOST_KUBECONFIG` cloning of minikube credentials for kubeconfig into devcontainer

## Issue reference

Resolves #3229
Resolves #2483

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [x] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#1520
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: docker/Dockerfile-dev
Comment (daixiang0) [inline docker/Dockerfile-dev:9]: Support bash is enough I think, we can introduce zsh if more need from users

**Review by daixiang0 — CHANGES_REQUESTED** [review]
Could you split pr? Seems it fixs too many issues.

File: docker/Dockerfile-dev
Comment (CodeMonkeyLeet) [reply docker/Dockerfile-dev:9]: Yeah, I don't feel strongly about this; it was just kept as the default setting that GitHub Codespaces use for their image. Any other folks want to weigh in? Given the support for dotfiles, users who want ZSH could add it as part of their install scripts.

**Comment by CodeMonkeyLeet** [comment]
> Could you split pr? Seems it fixs too many issues.

There's not much to split out:
- The bulk of it is making the devcontainer work in Codespaces, which introduces most of the changes. 
    - The two GitHub issues the PR resolves result from the same change to support Codespaces. The default Codespaces support for Docker is docker-in-docker, so the switch to support Codespaces resolves the request for docker-in-docker as well.
- The problem is that the existing devcontainer definition provides additional functionality for dealing with docker and kubectl when using local vscode devcontainers that is incompatible with Codespaces. 38cb75b addresses the regressions in that functionality by adding the support back in a way that is compatible with the Codespaces config. I would prefer not to split them apart because it would mean the first PR introduces functional regressions that a second PR is necessary to fix.
- 991a933 could be split out because it's really fixing an uncaught bug in the existing Devcontainer-dev definition, but because it's been refactored out into an entirely separate script introduced by the Codespaces work, I've included it so my testing of the scenario succeeds rather than resulting in an "expected" existing failure. 

Does that make sense?

**Comment by CodeMonkeyLeet** [comment]
@artursouza Can I get a review for this PR? @AaronCrawfis already merged the docs changes for this ahead of the code change and it would be nice if they were not out of sync with the code for a long time. 

A maintainer also needs to set the `DOCKER_REGISTRY` secret in GitHub for the dapr-dev-container workflow to succeed, and to publish the 0.1.3 version of the devcontainer to the daprio registry. A similar PR for components-contrib is blocked on that being available.

**Comment by CodeMonkeyLeet** [comment]
> Please, set it as a regular env variable value in the workflow yaml, unless it really contains a secret. The workflow already has ID and Password as secrets, so I don't foresee this being another secret.

@artur I pushed the change you suggested, but there's a tradeoff. It looks like `DAPR_REGISTRY` is defined as a GitHub repo secret not because it requires encryption, but because it's being used as a _portable runtime environment variable_ for GitHub Actions as described in https://github.com/dapr/dapr/blob/master/docs/development/setup-ci.md. Hardcoding it means that contributors who want to rely on the workflows to test changes in their fork prior to submission need to edit the workflows instead of just configuring the appropriate values in their forks as documented (i.e. this change breaks the docs).

**Comment by artursouza** [comment]
> > Please, set it as a regular env variable value in the workflow yaml, unless it really contains a secret. The workflow already has ID and Password as secrets, so I don't foresee this being another secret.
> 
> @artursouza I pushed the change you suggested, but there's a tradeoff. It looks like `DAPR_REGISTRY` is defined as a GitHub repo secret not because it requires encryption, but because it's being used as a _portable runtime environment variable_ for GitHub Actions as described in https://github.com/dapr/dapr/blob/master/docs/development/setup-ci.md. Hardcoding it means that contributors who want to rely on the workflows to test changes in their fork prior to submission need to edit the workflows instead of just configuring the appropriate values in their forks as documented (i.e. this change breaks the docs).

OK, so the secret for `DOCKER_REGISTRY: ${{ secrets.DOCKER_REGISTRY }}` already exists. The problem is that it cannot be used in a pull request workflow. Can you change the workflow not to use DOCKER_REGISTRY in a PR?

**Comment by CodeMonkeyLeet** [comment]
> OK, so the secret for `DOCKER_REGISTRY: ${{ secrets.DOCKER_REGISTRY }}` already exists. The problem is that it cannot be used in a pull request workflow. Can you change the workflow not to use DOCKER_REGISTRY in a PR?

Yes, I've updated the docker.mk and workflow to support that, thanks for the explanation!

**Review by artursouza — APPROVED** [review]

**Pull request merged**
